src/base/controller.hpp

```diff
@@ -137,6 +137,25 @@ class Controller {
                                  uint_t rng_seed,
                                  int num_threads_state) const = 0;
 
+  //-------------------------------------------------------------------------
+  // State validation
+  //-------------------------------------------------------------------------
+
+  // Return True if a given circuit (and internal noise model) are valid for
+  // execution on the given state.
+  template <class state_t>
+  static bool validate_state(const state_t &state,
```

Comment:
It's not used in this PR, but it is need in PR #13 (and in general for QasmSimulator to choose between multiple state types). I like the idea of just having one function and have a `bool throw_except = false` argument though

Reply:
For public APIs, is usually better to have two functions instead of default values, because is more expressive and clearly states intentions, otherwise you have to go to function definition to see what the bool actually means, and is more error prone.

Reply:
At least, use an `enum class` here instead of `bool`